toolchain/check/decl_name_stack.cpp

```diff
@@ -215,140 +236,71 @@ auto DeclNameStack::ApplyAndLookupName(NameContext& name_context,
-      auto& scope = context_->name_scopes().Get(scope_id);
-      if (scope.is_closed_import) {
-        CARBON_DIAGNOSTIC(QualifiedDeclOutsidePackage, Error,
-                          "Imported packages cannot be used for declarations.");
-        CARBON_DIAGNOSTIC(QualifiedDeclOutsidePackageSource, Note,
-                          "Package imported here.");
-        context_->emitter()
-            .Build(name_context.loc_id, QualifiedDeclOutsidePackage)
-            .Note(scope.inst_id, QualifiedDeclOutsidePackageSource)
-            .Emit();
-        // Only error once per package.
-        scope.is_closed_import = false;
-      }
-      if (!is_unqualified) {
-        PushNameQualifierScope(*context_, name_context.resolved_inst_id,
-                               scope_id,
-                               context_->name_scopes().Get(scope_id).has_error);
-      }
-      break;
-    }
-    default:
-      name_context.state = NameContext::State::ResolvedNonScope;
-      break;
+    name_context.state = NameContext::State::Resolved;
   }
 }
 
-auto DeclNameStack::CheckValidAsQualifier(const NameContext& name_context)
-    -> bool {
+auto DeclNameStack::ResolveAsScope(const NameContext& name_context,
+                                   const NameComponent& name) const
+    -> SemIR::NameScopeId {
   switch (name_context.state) {
-    case NameContext::State::Error:
-      // Already in an error state, so return without examining.
-      return false;
-
-    case NameContext::State::Resolved:
-      return true;
-
     case NameContext::State::Empty:
       CARBON_FATAL() << "No qualifier to resolve";
 
-    case NameContext::State::Finished:
-      CARBON_FATAL() << "Added a qualifier after calling FinishName";
+    case NameContext::State::Resolved:
+      break;
 
     case NameContext::State::Unresolved:
       // Because more qualifiers were found, we diagnose that the earlier
       // qualifier failed to resolve.
       context_->DiagnoseNameNotFound(name_context.loc_id,
                                      name_context.unresolved_name_id);
-      return false;
+      return SemIR::NameScopeId::Invalid;
 
-    case NameContext::State::ResolvedNonScope: {
-      // Because more qualifiers were found, we diagnose that the earlier
-      // qualifier didn't resolve to a scoped entity.
-      if (auto class_decl = context_->insts().TryGetAs<SemIR::ClassDecl>(
-              name_context.resolved_inst_id)) {
+    case NameContext::State::Finished:
+      CARBON_FATAL() << "Added a qualifier after calling FinishName";
+
+    case NameContext::State::Error:
+      // Already in an error state, so return without examining.
+      return SemIR::NameScopeId::Invalid;
+  }
+
+  auto new_params =
+      DeclParams(name.name_loc_id, name.implicit_params_id, name.params_id);
+
+  // Find the scope corresponding to the resolved instruction.
+  CARBON_KIND_SWITCH(context_->insts().Get(name_context.resolved_inst_id)) {
```

Comment:
Combining multiple functions like this is making it pretty long (110 LOC), is it worth thinking about how to keep it broken up a little? (partly I'm looking at it this way because it's now two switches)

You've previously suggested putting diagnostics in their own little helper functions, perhaps that would offer a few good cut points?

Reply:
Done.